File: caddisfly-app/app/src/androidTest/java/org/akvo/caddisfly/util/TestHelper.java
/*
 * Copyright (C) Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Caddisfly.
 *
 * Akvo Caddisfly is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Caddisfly is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Caddisfly. If not, see <http://www.gnu.org/licenses/>.
 */

package org.akvo.caddisfly.util;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;
import android.content.res.AssetManager;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.os.Build;
import android.os.Environment;
import android.preference.PreferenceManager;
import android.util.DisplayMetrics;

import androidx.annotation.StringRes;
import androidx.test.espresso.Espresso;
import androidx.test.espresso.NoMatchingViewException;
import androidx.test.espresso.ViewInteraction;
import androidx.test.rule.ActivityTestRule;
import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry;
import androidx.test.uiautomator.By;
import androidx.test.uiautomator.UiDevice;
import androidx.test.uiautomator.UiObject;
import androidx.test.uiautomator.UiObject2;
import androidx.test.uiautomator.UiObjectNotFoundException;
import androidx.test.uiautomator.UiScrollable;
import androidx.test.uiautomator.UiSelector;

import org.akvo.caddisfly.R;
import org.akvo.caddisfly.app.CaddisflyApp;
import org.akvo.caddisfly.common.AppConfig;
import org.akvo.caddisfly.common.TestConstants;

import java.io.File;
import java.util.Collection;
import java.util.HashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

import timber.log.Timber;

import static androidx.test.espresso.Espresso.onView;
import static androidx.test.espresso.action.ViewActions.click;
import static androidx.test.espresso.assertion.ViewAssertions.matches;
import static androidx.test.espresso.matcher.ViewMatchers.isDisplayed;
import static androidx.test.espresso.matcher.ViewMatchers.withContentDescription;
import static androidx.test.espresso.matcher.ViewMatchers.withId;
import static androidx.test.espresso.matcher.ViewMatchers.withParent;
import static androidx.test.espresso.matcher.ViewMatchers.withText;
import static androidx.test.platform.app.InstrumentationRegistry.getInstrumentation;
import static androidx.test.runner.lifecycle.Stage.RESUMED;
import static org.akvo.caddisfly.util.TestUtil.clickListViewItem;
import static org.akvo.caddisfly.util.TestUtil.findButtonInScrollable;
import static org.akvo.caddisfly.util.TestUtil.sleep;
import static org.hamcrest.Matchers.allOf;

public final class TestHelper {

    public static final String mCurrentLanguage = "en";
    private static final boolean TAKE_SCREENSHOTS = true;
    private static final Map<String, String> STRING_HASH_MAP_EN = new HashMap<>();
    private static final Map<String, String> STRING_HASH_MAP_ES = new HashMap<>();
    private static final Map<String, String> STRING_HASH_MAP_FR = new HashMap<>();
    private static final Map<String, String> STRING_HASH_MAP_IN = new HashMap<>();
    public static Map<String, String> currentHashMap;
    public static UiDevice mDevice;

    private static int mCounter;

    private TestHelper() {
    }

    private static void addString(String key, String... values) {
        STRING_HASH_MAP_EN.put(key, values[0]);
        if (values.length > 1) {
            STRING_HASH_MAP_ES.put(key, values[1]);
            STRING_HASH_MAP_FR.put(key, values[2]);
            STRING_HASH_MAP_IN.put(key, values[3]);
        } else {
            STRING_HASH_MAP_ES.put(key, values[0]);
            STRING_HASH_MAP_FR.put(key, values[0]);
            STRING_HASH_MAP_IN.put(key, values[0]);
        }
    }

    public static String getString(@StringRes int resourceId) {
        return getString(getCurrentActivity(), resourceId);
    }

    public static String getString(Activity activity, @StringRes int resourceId) {
        Resources currentResources = activity.getResources();
        AssetManager assets = currentResources.getAssets();
        DisplayMetrics metrics = currentResources.getDisplayMetrics();
        Configuration config = new Configuration(currentResources.getConfiguration());
        config.locale = new Locale(mCurrentLanguage);
        Resources res = new Resources(assets, metrics, config);

        return res.getString(resourceId);
    }

    public static void loadData(Activity activity, String languageCode) {

        STRING_HASH_MAP_EN.clear();
        STRING_HASH_MAP_ES.clear();
        STRING_HASH_MAP_FR.clear();
        STRING_HASH_MAP_IN.clear();

        Resources currentResources = activity.getResources();
        AssetManager assets = currentResources.getAssets();
        DisplayMetrics metrics = currentResources.getDisplayMetrics();
        Configuration config = new Configuration(currentResources.getConfiguration());
        config.locale = new Locale(languageCode);
        Resources res = new Resources(assets, metrics, config);

        addString(TestConstant.LANGUAGE, "English", "Español", "Français", "Bahasa Indonesia");
//        addString("otherLanguage", "Français", "English");
        addString(TestConstant.FLUORIDE, res.getString(R.string.fluoride));
        addString("chlorine", res.getString(R.string.freeChlorine));
        addString("survey", res.getString(R.string.survey));
        addString("sensors", res.getString(R.string.sensors));
        addString("electricalConductivity", res.getString(R.string.electricalConductivity));
        addString("unnamedDataPoint", res.getString(R.string.unnamedDataPoint));
        addString("createNewDataPoint", res.getString(R.string.addDataPoint));
        addString(TestConstant.USE_EXTERNAL_SOURCE, res.getString(R.string.useExternalSource));
        addString(TestConstant.GO_TO_TEST, res.getString(R.string.goToTest));
        addString("next", res.getString(R.string.next));

        // Restore device-specific locale
        new Resources(assets, metrics, currentResources.getConfiguration());

        switch (languageCode) {
            case "en":
                currentHashMap = STRING_HASH_MAP_EN;
                break;
            case "es":
                currentHashMap = STRING_HASH_MAP_ES;
                break;
            case "in":
                currentHashMap = STRING_HASH_MAP_IN;
                break;
            default:
                currentHashMap = STRING_HASH_MAP_FR;
                break;
        }
    }

    public static void takeScreenshot() {
        if (TAKE_SCREENSHOTS && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            File path = new File(Environment.getExternalStorageDirectory().getPath()
                    + "/Akvo Caddisfly/screenshots/screen-" + mCounter++ + "-" + mCurrentLanguage + ".png");
            mDevice.takeScreenshot(path, 0.5f, 60);
        }
    }

    public static void takeScreenshot(String name, int page) {
        if (TAKE_SCREENSHOTS && Build.VERSION.SDK_INT >= Build.VERSION_CODES.JELLY_BEAN_MR1) {
            File path;
            path = new File(Environment.getExternalStorageDirectory().getPath()
                    + "/Akvo Caddisfly/screenshots/" + name + "-" + mCurrentLanguage + "-" +
                    String.format("%02d", page + 1) + ".png");
            mDevice.takeScreenshot(path, 0.1f, 30);
        }
    }

    public static void goToMainScreen() {

        boolean found = false;
        while (!found) {
            try {
                onView(withId(R.id.button_info)).check(matches(isDisplayed()));
                found = true;
            } catch (NoMatchingViewException e) {
                Espresso.pressBack();
            }
        }
    }

    public static void activateTestMode() {

        if (!AppConfig.IS_TEST_MODE) {

            onView(withId(R.id.button_info)).perform(click());

            onView(withText(R.string.about)).check(matches(isDisplayed())).perform(click());

            String version = CaddisflyApp.getAppVersion(false);

            onView(withText(version)).check(matches(isDisplayed()));

            enterDiagnosticMode();

            onView(withId(R.id.actionSettings)).perform(click());

            clickListViewItem("Test Mode");
        }

        goToMainScreen();
    }

    public static void clickExternalSourceButton(int index) {

        String buttonText = currentHashMap.get(TestConstant.GO_TO_TEST);

        if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            assert buttonText != null;
            buttonText = buttonText.toUpperCase();
        }

        findButtonInScrollable(buttonText);

        List<UiObject2> buttons = mDevice.findObjects(By.text(buttonText));

        if (index < buttons.size()) {
            buttons.get(index).click();
        } else {
            UiScrollable listView = new UiScrollable(new UiSelector());
            try {
                listView.scrollToEnd(1);
            } catch (UiObjectNotFoundException e) {
                e.printStackTrace();
            }

            List<UiObject2> buttons1 = mDevice.findObjects(By.text(buttonText));
            buttons1.get(buttons1.size() - 1).click();
        }

        // New Android OS seems to popup a button for external app
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.M) {
            sleep(1000);
            mDevice.findObject(By.text("Akvo Caddisfly")).click();
            sleep(1000);
        }

        mDevice.waitForWindowUpdate("", 2000);

        sleep(4000);
    }

    public static void clickExternalSourceButton(String text) {
        try {

            String buttonText = currentHashMap.get(text);

            if (android.os.Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                assert buttonText != null;
                buttonText = buttonText.toUpperCase();
            }

            findButtonInScrollable(buttonText);

            assert buttonText != null;
            mDevice.findObject(new UiSelector().text(buttonText)).click();

            // New Android OS seems to popup a button for external app
            if (android.os.Build.VERSION.SDK_INT == Build.VERSION_CODES.M
                    && (text.equals(TestConstant.USE_EXTERNAL_SOURCE)
                    || text.equals(TestConstant.GO_TO_TEST))) {
                sleep(1000);
                mDevice.findObject(By.text("Akvo Caddisfly")).click();
                sleep(1000);
            }

            mDevice.waitForWindowUpdate("", 2000);

        } catch (UiObjectNotFoundException e) {
            Timber.e(e);
        }
    }

    public static void gotoSurveyForm() {

        Context context = getInstrumentation().getContext();
        Intent intent = context.getPackageManager().getLaunchIntentForPackage(TestConstants.FLOW_SURVEY_PACKAGE_NAME);
        if (intent != null) {
            intent.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK);
        }
        context.startActivity(intent);

        if (!clickListViewItem(currentHashMap.get("unnamedDataPoint"))) {

            UiObject addButton = mDevice.findObject(new UiSelector()
                    .resourceId("org.akvo.flow:id/add_data_point_fab"));

            try {
                if (addButton.exists() && addButton.isEnabled()) {
                    addButton.click();
                }
            } catch (UiObjectNotFoundException e) {
                Timber.e(e);
            }
        }

        // mDevice.findObject(By.text("Caddisfly Tests")).click();
    }

    public static void enterDiagnosticMode() {
        for (int i = 0; i < 10; i++) {
            onView(withId(R.id.textVersion)).perform(click());
        }
    }

    public static void leaveDiagnosticMode() {
        onView(withId(R.id.fabDisableDiagnostics)).perform(click());
    }

//    public static void resetLanguage() {
//
//        goToMainScreen();
//
//        onView(withId(R.id.actionSettings)).perform(click());
//
//        onView(withText(R.string.language)).perform(click());
//
//        onData(Matchers.hasToString(Matchers.startsWith(currentHashMap.get("language")))).perform(click());
//
//        mDevice.waitForIdle();
//
//        sleep(5000);
//
//        mDevice.waitForIdle();
//
//        Espresso.pressBack();
//
//        onView(withId(R.id.actionSettings)).perform(click());
//
//        onView(withText(R.string.language)).perform(click());
//
//        onData(Matchers.hasToString(Matchers.startsWith(currentHashMap.get("language")))).perform(click());
//
//        mDevice.waitForIdle();
//
//        sleep(5000);
//
//        mDevice.waitForIdle();
//
//        goToMainScreen();
//
//    }

    public static void clearPreferences(ActivityTestRule activityTestRule) {
        SharedPreferences prefs =
                PreferenceManager.getDefaultSharedPreferences(activityTestRule.getActivity());
        prefs.edit().clear().apply();
    }

    public static void navigateUp() {
        ViewInteraction appCompatImageButton = onView(
                allOf(withContentDescription(R.string.navigate_up),
                        withParent(withId(R.id.toolbar)),
                        isDisplayed()));
        appCompatImageButton.perform(click());
    }

    public static Activity getCurrentActivity() {
        final Activity[] currentActivity = {null};
        getInstrumentation().runOnMainSync(() -> {
            Collection resumedActivities = ActivityLifecycleMonitorRegistry.getInstance()
                    .getActivitiesInStage(RESUMED);
            if (resumedActivities.iterator().hasNext()) {
                currentActivity[0] = (Activity) resumedActivities.iterator().next();
            }
        });
        return currentActivity[0];
    }

    public static boolean isLowMemoryDevice(String model) {
        return model.contains("ASUS_Z007");
    }
}

File: caddisfly-app/app/src/androidTest/java/org/akvo/caddisfly/util/TestHelper.kt
/*
 * Copyright (C) Stichting Akvo (Akvo Foundation)
 *
 * This file is part of Akvo Caddisfly.
 *
 * Akvo Caddisfly is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Akvo Caddisfly is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Akvo Caddisfly. If not, see <http://www.gnu.org/licenses/>.
 */

@file:Suppress("DEPRECATION")

package org.akvo.caddisfly.util

import android.app.Activity
import android.content.Intent
import android.content.res.Configuration
import android.content.res.Resources
import android.graphics.drawable.ColorDrawable
import android.os.Build
import android.preference.PreferenceManager
import android.provider.Settings
import android.view.View
import androidx.annotation.StringRes
import androidx.core.content.ContextCompat
import androidx.test.espresso.Espresso
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.NoMatchingViewException
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.*
import androidx.test.platform.app.InstrumentationRegistry.getInstrumentation
import androidx.test.rule.ActivityTestRule
import androidx.test.runner.lifecycle.ActivityLifecycleMonitorRegistry
import androidx.test.runner.lifecycle.Stage.RESUMED
import androidx.test.uiautomator.*
import org.akvo.caddisfly.BuildConfig
import org.akvo.caddisfly.R
import org.akvo.caddisfly.common.AppConstants.FLOW_SURVEY_PACKAGE_NAME
import org.akvo.caddisfly.helper.FileHelper
import org.akvo.caddisfly.helper.FileHelper.getUnitTestImagesFolder
import org.akvo.caddisfly.util.TestUtil.childAtPosition
import org.hamcrest.Matchers.`is`
import org.hamcrest.Matchers.allOf
import org.junit.Assert
import timber.log.Timber
import java.io.File
import java.util.*

lateinit var mDevice: UiDevice

fun isStripPatchAvailable(name: String = "."): Boolean {
    val file = File(getUnitTestImagesFolder(), "$name.yuv")
    return file.exists()
}

/**
 * Skip opening external Survey if the tests are running in Firebase Test Lab
 * or if the phone OS is too old
 */
fun skipOpeningExternalApp(model: Int = Build.VERSION_CODES.LOLLIPOP): Boolean {

    try {
        val testLabSetting: String? = Settings.System.getString(
                getInstrumentation().targetContext.contentResolver, "firebase.test.lab")

        if ("true" == testLabSetting) {
            return true
        }
    } catch (e: IllegalStateException) {
    }

    return model < Build.VERSION_CODES.LOLLIPOP
}

object TestHelper {

    private val STRING_HASH_MAP_EN = HashMap<String, String>()
    private val STRING_HASH_MAP_ES = HashMap<String, String>()
    private val STRING_HASH_MAP_FR = HashMap<String, String>()
    private val STRING_HASH_MAP_IN = HashMap<String, String>()
    private lateinit var currentHashMap: Map<String, String>

    private var screenshotCount = -1

    val currentActivity: Activity
        get() {
            val currentActivity = arrayOf<Any>({ null })
            getInstrumentation().runOnMainSync {
                val resumedActivities = ActivityLifecycleMonitorRegistry.getInstance()
                        .getActivitiesInStage(RESUMED)
                if (resumedActivities.iterator().hasNext()) {
                    currentActivity[0] = resumedActivities.iterator().next() as Activity
                }
            }
            return currentActivity[0] as Activity
        }


    private fun addString(key: String, vararg values: String) {
        STRING_HASH_MAP_EN[key] = values[0]
        if (values.size > 1) {
            STRING_HASH_MAP_ES[key] = values[1]
            STRING_HASH_MAP_FR[key] = values[2]
            STRING_HASH_MAP_IN[key] = values[3]
        } else {
            STRING_HASH_MAP_ES[key] = values[0]
            STRING_HASH_MAP_FR[key] = values[0]
            STRING_HASH_MAP_IN[key] = values[0]
        }
    }

    fun getString(@StringRes resourceId: Int): String {
        return getString(currentActivity, resourceId)
    }

    fun getString(activity: Activity, @StringRes resourceId: Int): String {
        val currentResources = activity.resources
        val assets = currentResources.assets
        val metrics = currentResources.displayMetrics
        val config = Configuration(currentResources.configuration)
        config.locale = Locale(BuildConfig.TEST_LANGUAGE)
        val res = Resources(assets, metrics, config)

        return res.getString(resourceId)
    }

    fun loadData(activity: Activity, languageCode: String) {

        STRING_HASH_MAP_EN.clear()
        STRING_HASH_MAP_ES.clear()
        STRING_HASH_MAP_FR.clear()
        STRING_HASH_MAP_IN.clear()

        val currentResources = activity.resources
        val assets = currentResources.assets
        val metrics = currentResources.displayMetrics
        val config = Configuration(currentResources.configuration)
        config.locale = Locale(languageCode)
        val res = Resources(assets, metrics, config)

        addString(TestConstant.LANGUAGE, "English", "Español", "Français", "Bahasa Indonesia")
        //        addString("otherLanguage", "Français", "English");
        addString(TestConstant.FLUORIDE, res.getString(R.string.fluoride))
        addString("chlorine", res.getString(R.string.freeChlorine))
        addString("survey", res.getString(R.string.survey))
        addString("sensors", res.getString(R.string.sensors))
        addString("electricalConductivity", res.getString(R.string.electricalConductivity))
        addString("unnamedDataPoint", res.getString(R.string.unnamedDataPoint))
        addString("createNewDataPoint", res.getString(R.string.addDataPoint))
        addString(TestConstant.USE_EXTERNAL_SOURCE, res.getString(R.string.useExternalSource))
        addString(TestConstant.GO_TO_TEST, res.getString(R.string.goToTest))
        addString("next", res.getString(R.string.next))

        // Restore device-specific locale
        Resources(assets, metrics, currentResources.configuration)

        currentHashMap = when (languageCode) {
            "en" -> STRING_HASH_MAP_EN
            "es" -> STRING_HASH_MAP_ES
            "in" -> STRING_HASH_MAP_IN
            else -> STRING_HASH_MAP_FR
        }
    }

    fun takeScreenshot() {
        takeScreenshot("app", screenshotCount++)
    }

    fun takeScreenshot(name: String, page: Int) {
        if (BuildConfig.TAKE_SCREENSHOTS) {
            val folder = FileHelper.getScreenshotFolder()
            val path = File(folder, name + "-" + BuildConfig.TEST_LANGUAGE + "-" +
                    String.format("%02d", page + 1) + ".png")

            if (!folder.exists()) {
                folder.mkdirs()
            }

            mDevice.takeScreenshot(path, 0.1f, 30)
        }
    }

    fun goToMainScreen() {

        var found = false
        while (!found) {
            try {
                onView(withId(R.id.button_info)).check(matches(isDisplayed()))
                found = true
            } catch (e: NoMatchingViewException) {
                Espresso.pressBack()
            }

        }
    }

    fun activateTestMode() {

//        @Suppress("ConstantConditionIf")
//        if (!IS_TEST_MODE) {
//
//            onView(withId(R.id.button_info)).perform(click())
//
//            onView(withText(R.string.about)).check(matches(isDisplayed())).perform(click())
//
//            val version = CaddisflyApp.getAppVersion(false)
//
//            onView(withText(version)).check(matches(isDisplayed()))
//
//            enterDiagnosticMode()
//
//            onView(withId(R.id.actionSettings)).perform(click())
//
//            clickListViewItem("Test Mode")
//        }

        goToMainScreen()
    }

    fun clickExternalSourceButton(index: Int) {

        var buttonText = currentHashMap[TestConstant.GO_TO_TEST]

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
            assert(buttonText != null)
            buttonText = buttonText!!.toUpperCase()
        }

        findButtonInScrollable(buttonText!!)

        val buttons = mDevice.findObjects(By.text(buttonText))

        if (index < buttons.size) {
            buttons[index].click()
        } else {
            val listView = UiScrollable(UiSelector())
            try {
                listView.scrollToEnd(1)
            } catch (e: UiObjectNotFoundException) {
                e.printStackTrace()
            }

            val buttons1 = mDevice.findObjects(By.text(buttonText))
            buttons1[buttons1.size - 1].click()
        }

        // New Android OS seems to popup a button for external app
        if (Build.VERSION.SDK_INT == Build.VERSION_CODES.M) {
            sleep(1000)
            mDevice.findObject(By.text("Akvo Caddisfly")).click()
            sleep(1000)
        }

        mDevice.waitForWindowUpdate("", 2000)

        sleep(4000)
    }

    fun clickExternalSourceButton(text: String) {
        try {

            var buttonText = currentHashMap[text]

            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.N) {
                assert(buttonText != null)
                buttonText = buttonText!!.toUpperCase()
            }

            findButtonInScrollable(buttonText!!)

            mDevice.findObject(UiSelector().text(buttonText)).click()

            // New Android OS seems to popup a button for external app
            if (Build.VERSION.SDK_INT == Build.VERSION_CODES.M && (text == TestConstant.USE_EXTERNAL_SOURCE || text == TestConstant.GO_TO_TEST)) {
                sleep(1000)
                mDevice.findObject(By.text("Akvo Caddisfly")).click()
                sleep(1000)
            }

            mDevice.waitForWindowUpdate("", 2000)

        } catch (e: UiObjectNotFoundException) {
            Timber.e(e)
        }
    }

    fun gotoSurveyForm() {

        val context = getInstrumentation().context
        val intent = context.packageManager.getLaunchIntentForPackage(FLOW_SURVEY_PACKAGE_NAME)
        intent?.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TASK)
        context.startActivity(intent)

        if (!currentHashMap["unnamedDataPoint"]?.let { clickListViewItem(it) }!!) {

            val addButton = mDevice.findObject(UiSelector()
                    .resourceId("org.akvo.flow:id/add_data_point_fab"))

            try {
                if (addButton.exists() && addButton.isEnabled) {
                    addButton.click()
                }
            } catch (e: UiObjectNotFoundException) {
                Timber.e(e)
            }

        }
    }

    fun enterDiagnosticMode() {
        for (i in 0..9) {
            onView(withId(R.id.textVersion)).perform(click())
        }
    }

    fun leaveDiagnosticMode() {
        onView(withId(R.id.fabDisableDiagnostics)).perform(click())
    }

    fun clearPreferences(activityTestRule: ActivityTestRule<*>) {
        val prefs = PreferenceManager.getDefaultSharedPreferences(activityTestRule.activity)
        prefs.edit().clear().apply()
    }

    fun navigateUp() {
        onView(allOf(withContentDescription(R.string.navigate_up),
                isDisplayed())).perform(click())
    }

    fun isDeviceInitialized(): Boolean {
        return ::mDevice.isInitialized
    }

    fun clickSubmitButton() {
        onView(allOf(withId(R.id.buttonSubmit), withText(R.string.submitResult),
                childAtPosition(
                        allOf(withId(R.id.buttonsLayout),
                                childAtPosition(
                                        withClassName(`is`("android.widget.RelativeLayout")),
                                        2)),
                        0),
                isDisplayed())).perform(click())
    }

    fun clickCloseButton() {
        sleep(2000)
        try {
            onView(allOf(withId(R.id.buttonClose), withText(R.string.close),
                    isDisplayed())).perform(click())
        } catch (e: Exception) {
            Timber.e(e)
        }
        sleep(2000)
    }

    fun clickStartButton() {
        onView(allOf(withId(R.id.buttonStart), withText(R.string.start),
                isDisplayed())).perform(click())
    }

    fun clickSubmitResultButton() {
        onView(allOf(withId(R.id.buttonSubmitResult), withText(R.string.submitResult), isDisplayed()))
                .perform(click())
    }

    fun assertBackgroundColor(view: Int, color: Int) {
        val bar = currentActivity.findViewById<View>(view)
        val actualColor = (bar.background as ColorDrawable).color
        val expectedColor = ContextCompat.getColor(currentActivity, color)
        Assert.assertEquals(actualColor, expectedColor)
    }
}
